File: python/str/kmp.py
# Copyright [2018] [Haibo Yan]
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

class KMP:
    def kmp(self, s, p):
        lp = self.pattern_index(p)

        k = 0
        for si, sc in enumerate(s):
            if sc == p[k]:
                k = k + 1
            else:
                k = lp[k]
            if k == len(p):
                return si - k + 1

        return -1


    def pattern_index(self, p):
        if not p:
            return []

        indexes = [0]
        k = 0
        for i in range(1, len(p)):
            c = p[i]
            while k > 0 and p[k] != c:
                k = indexes[k - 1]

            if p[k] == c:
                k = k + 1

            indexes.append(k)
        return indexes

File: c++/algorithm/str/KMP.cpp
//Copyright [2018] [Haibo Yan]
//
//Licensed under the Apache License, Version 2.0 (the "License");
//you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.

//
// Created by haibo on 3/17/18.
//

#include "KMP.h"

int* KMP::pattern(string pattern) {
    int *index = new int[pattern.size()];

    index[0] = 0;

    int i = 0;

    for (int j = 1; j < pattern.size(); j++) {
        while (i > 0 && pattern[j] != pattern[i]) {
            j = index[i];
        }
        if (pattern[i] == pattern[j]) {
            i++;
        }
        index[j] = i;
    }

    return index;
}

int KMP::substr(string s, string p) {
    int *pindex = pattern(p);

    int i = 0, j = 0;

    while (i < s.length() && j < p.length()) {
        if (s[i] == p[j]) {
            i++;
            j++;
        } else if (j > 0) {
            j = pindex[j - 1];
        } else {
            i++;
        }
    }

    if (j < p.length()) {
        return -1;
    } else {
        return i - j;
    }
}

